src/main/java/org/spongepowered/common/world/DimensionManager.java

```diff
@@ -24,336 +24,982 @@
+        final WorldServer worldServer = (WorldServer) new WorldServer(server, saveHandler, worldInfo, dimensionId, server.theProfiler)
+                .init();
+
+        // WorldSettings is only non-null here if this is a newly generated WorldInfo and therefore we need to initialize to calculate spawn.
+        if (worldSettings != null) {
+            worldServer.initialize(worldSettings);
+        }
+
+        worldServer.addEventListener(new WorldManager(server, worldServer));
+
+        // This code changes from Mojang's to account for per-world API-set GameModes.
+        if (!server.isSinglePlayer() && worldServer.getWorldInfo().getGameType().equals(WorldSettings.GameType.NOT_SET)) {
+            worldServer.getWorldInfo().setGameType(server.getGameType());
+        }
+
+        SpongeImpl.postEvent(SpongeImplHooks.createLoadWorldEvent((org.spongepowered.api.world.World) worldServer));
+
+        if (prepareSpawn) {
+            ((IMixinMinecraftServer) server).prepareSpawnArea(worldServer);
+        }
+        return worldServer;
     }
 
-    public static void registerDimension(int id, int providerType) {
-        if (!providers.containsKey(providerType)) {
-            throw new IllegalArgumentException(
-                    String.format("Failed to register dimension for id %d, provider type %d does not exist", id, providerType));
+    private static WorldServer[] reorderWorldsVanillaFirst() {
+        final List<WorldServer> sorted = new ArrayList<>(worldByDimensionId.valueCollection());
+
+        // Ensure that we'll load in Vanilla order then plugins
+        WorldServer worldServer = worldByDimensionId.get(0);
+        sorted.remove(worldServer);
+        sorted.add(0, worldServer);
+        worldServer = worldByDimensionId.get(-1);
+        if (worldServer != null) {
+            sorted.remove(worldServer);
+            sorted.add(1, worldServer);
         }
-        if (dimensions.containsKey(id)) {
-            throw new IllegalArgumentException(String.format("Failed to register dimension for id %d, One is already registered", id));
+        worldServer = worldByDimensionId.get(1);
+        if (worldServer != null) {
+            sorted.remove(worldServer);
+            sorted.add(2, worldServer);
         }
-        dimensions.put(id, providerType);
-        if (id >= 0) {
-            dimensionMap.set(id);
+        return sorted.toArray(new WorldServer[sorted.size()]);
+    }
+
+    /**
+     * Parses a {@link UUID} from disk from other known plugin platforms and sets it on the
+     * {@link WorldProperties}. Currently only Bukkit is supported.
+     */
+    private static UUID setUuidOnProperties(Path savesRoot, WorldProperties properties) {
+        checkNotNull(properties);
+
+        UUID uuid;
+        if (properties.getUniqueId() == null || properties.getUniqueId().equals
+                (UUID.fromString("00000000-0000-0000-0000-000000000000"))) {
```

Comment:
It would be a good idea to store this in a constant (like `EMPTY_UUID`). Also, is this the same as `new UUID(0, 0)`?